src/core/thread/mle.cpp

```diff
@@ -4741,5 +4759,58 @@ void Mle::DelayedResponseMetadata::RemoveFrom(Message &aMessage) const
     SuccessOrAssert(aMessage.SetLength(aMessage.GetLength() - sizeof(*this)));
 }
 
+Error Mle::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    OT_ASSERT(IsChild());
+
+    if (mDetachGracefullyTimer.IsRunning())
+    {
+        ExitNow(error = kErrorBusy);
+    }
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    mDetachGracefullyTimer.Start(kDetachGracefullyTimeout);
+
+    IgnoreError(SendChildUpdateRequest(/*aTimeout=*/0));
+
+exit:
+    return error;
+}
+
+void Mle::HandleDetachGracefullyTimer(Timer &aTimer)
+{
+    aTimer.Get<Mle>().HandleDetachGracefullyTimer();
+}
+
+void Mle::HandleDetachGracefullyTimer(void)
+{
+    Stop();
```

Comment:
At this moment, the timer is already stopped. 

It's possible that `DetachGracefully` might be called again before the tasklet is executed, which would think that there is no gradeful detaching ongoing, but it's wrong.

I think we need to make sure the check condition for "whether graceful detaching is ongoing" to hold true during the whole process.

Reply:
You are right. Stop() needs to synchronously call the callback. So I made the "task posting" between "detach gracefully" and Stop(), instead of between Stop() and the callback.

I.e. from `DetachGracefully -> Stop -> Post -> callback` to `DetachGracefully -> Post -> Stop -> callback`.

When the callback is called, the state must be disabled.

Reply:
Calling `Stop()` directly here looks good to me.

The timer callback has already guaranteed that we are not calling the callback synchronously.